### PR TITLE
Add guide about testing a custom transaction - Closes #485

### LINKED ISSUE #485: New guide: Test a transaction
Write a guide how to test the logic of s custom transaction with unit tests.

### PR DESCRIPTION
**Pull request by Tschakki**
Closes #485

### CODE REVIEW

File: modules/ROOT/pages/guides/test.adoc [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+:url_wiki_stubs: https://en.wikipedia.org/wiki/Test_stub
+
+:url_tutorials_transport_3: tutorials/transport2.adoc#register_packet
+
+The more complex the logic inside the custom transaction types, the more complicated it becomes to verify that the custom transaction logic is working as expected.
+
+Therefore it is recommended to write {url_unit_tests}[unit tests], that verify the logic of the transaction type.
+
+Especially for verifying the code of the `undoAsset()` function, it is convenient to write unit tests.
+This is due to the fact that the code in the `undoAsset` function is only executed, if the node discovers itself on a fork with the main chain.
```
Comment (michielmulders) [inline modules/ROOT/pages/guides/test.adoc]: Really like the introduction! 
Very small comment, maybe remove the comma here: `is only executed, if the` ?

File: modules/ROOT/pages/guides/test.adoc [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+[NOTE]
+====
+To be on a **fork** means that the node added some different blocks to the chain than its peers.
+In order to sync again with the network, it has to remove the blocks that are different, and undo the transactions inside these blocks.
+To undo the transaction, the `undoAsset()` function will be called for each transaction inside of the blocks that need to be discarded.
```
Comment (michielmulders) [inline modules/ROOT/pages/guides/test.adoc]: What about changing `discarded` to `reverted`?

File: modules/ROOT/pages/guides/test.adoc [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+====
+
+To explain how to write unit tests for custom transactions, we will use the example of the `undoAsset()` method of the xref:{url_tutorials_transport_3}[supply chain tutorial: RegisterPacketTransaction].
+You can use it as template to write unit tests for your own custom transactions.
+
+NOTE: To test the complete logic of a custom transaction, it is required to write unit tests for all methods that are implemented in the custom transactions: `validateAsset()`, `applyAsset()` and `undoAsset`.
```
Comment (michielmulders) [inline modules/ROOT/pages/guides/test.adoc]: For consistency, add the `()` brackets to `undoAsset` as well at the end of the sentence.

File: modules/ROOT/pages/guides/test.adoc [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+== Write a test script
+
+First, let's look at the `undoAsset()` function of the {url_github_transport_register_packet}[RegisterPacketTransaction].
+
+.Snippet of undoAsset(), File: transactions/solutions/register-packet.js
+[source,js]
+----
+undoAsset(store) {
```
Comment (michielmulders) [inline modules/ROOT/pages/guides/test.adoc:39]: On [staging](https://liskhq.github.io/lisk-docs/lisk-sdk/guides/test.html#_write_a_test_script) this snippet has no markup while the one below (transactions/tests/register-packet.test.js) does have code markup? Both have `[source,js]` tags tho?

File: modules/ROOT/pages/guides/test.adoc [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    /* The postage is removed from the packet account balance, and packet data in the asset is nullified */
+    const packet = store.account.get(this.asset.packetId);
+    const originalPacketAccount = { ...packet, balance: 0, asset: null };
+    store.account.set(packet.address, originalPacketAccount);
+
+    return errors;
+}
+----
+
+To test, that the `undoAsset()` function is working as expected, write a unit test {url_github_transport_test_register}[register-packet.test.js]:
```
Comment (michielmulders) [inline modules/ROOT/pages/guides/test.adoc]: Would it make sense to mention explicitly that we are using `jest` for testing? And we also use extra helper module `jest-when` for being able to mock calls more easily.

File: modules/ROOT/pages/guides/test.adoc [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+----
+
+<1> The `get()` and `set()` functions of the `stateStore` are mocked by creating {url_wiki_stubs}[stubs] in the `beforeEach()` function.
+This allows replace the call with a custom return value.
+<2> Now start the test, add a short and precise description of what the test is about.
+<3> When `storeStub.account.get` is called with `asset.senderId`, we replace the return value with the `mockedSenderAccount`.
+<4> When `storeStub.account.get` is called with `asset.packetId`, we replace the return value with the `mockedPacketAccount`.
+<5> A new transaction is created.
+<6> The `undoAsset()` function of the transaction is called, and the previousely defined `storeStub` is passed to the `undoAsset()` function.
+<7> Here start actually the tests.
```
Comment (michielmulders) [inline modules/ROOT/pages/guides/test.adoc]: Maybe flip the order of words to: The tests start here actually?

File: modules/ROOT/pages/guides/test.adoc (hunk 5 above)
Comment (michielmulders) [reply modules/ROOT/pages/guides/test.adoc]: I see that we mention to install Jest a bit more below in the `run the test script`. But maybe to give them already an idea of what syntax we use, we can mention that the tests are written in Jest here.

File: modules/ROOT/pages/guides/test.adoc [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+....
+
+In the example above, all expectations were met and the test passed.
+
+== What else needs to be tested?
+Is writing unit tests really enough to ensure the functionality of a custom transaction type?
+
+**Short answer: The unit tests are sufficient.**
+
+**Explanation:** You may wonder if it is required to write additional functional and integration tests.
```
Comment (michielmulders) [inline modules/ROOT/pages/guides/test.adoc:222]: Helpful info!

**Review by michielmulders — CHANGES_REQUESTED** [review]

File: modules/ROOT/pages/guides/test.adoc (hunk 4 above)
Comment (Tschakki) [reply modules/ROOT/pages/guides/test.adoc:39]: they both have syntax highlighting, but first snippet just doesnt have strings ^^

**Review by michielmulders — APPROVED** [review]

**Pull request merged**